[dv] Enable verification of the Bitmanip Extension with OVPsim and Spike

vogelpi opened the pull request:
This PR enables verification of the Bitmanip Extension with OVPsim and Spike and adds a new test that can be updated in the future to also include newly added instruction groups. For Spike,
a different branch has to be used. This PR updates the documentation accordingly.

What is still missing is support for coverage. This will be later vendored in from upstream RISCV-DV.

This is related to lowRISC/ibex#703.

vogelpi commented:
I guess I will also have to update CI (to use the Bitmanip-enabled version of Spike) + I would need someone to actually check OVPsim.

vogelpi commented:
Pinging also @weicaiyang as he is working on the coverage.

vogelpi commented:
Ah yes, we don't the UVM tests in CI here. I mixed that up. I also thought the instructions could be cleaned up.

The Zbb and Zbt tests passed, but I just ran a couple of iterations. Let me check if also the other tests are still passing.

udinator commented:
Thanks for doing this Pirmin! I'll try this out with OVPsim on my end today to check that it works.

weicaiyang commented:
Thanks Pirmin. Coverage will be completed soon.

vogelpi commented:
@ganoam noted that only the first option `enable_bitmanip_groups` in `testlist.yaml` is taken into account. To enable multiple groups, they have to be provided as a comma separated list:
```
+enable_bitmanip_groups=zbb,zbt
``` 
I have now adapted the testlist accordingly. We get some failures in Zbt now. I suggest to debug these once this PR is in.

ganoam commented:
Oh. thanks. I am already on it.

vogelpi commented:
As for the other tests, we get failures in:
- riscv_mmu_stress_test
- riscv_debug_branch_jump_test

udinator commented:
> As for the other tests, we get failures in:
> 
> * riscv_mmu_stress_test

There are two underlying issues causing failures here, one was an RTL bug that Tom has pushed in a fix for, the second is a more subtle but annoying problem that I have a large patch in riscv-dv to fix, waiting for a review on that end.

> * riscv_debug_branch_jump_test

This is an issue with the debug mode sub-program generation in riscv-dv as well that I need to root cause.

udinator commented on dv/uvm/core_ibex/riscv_dv_extension/riscvOVPsim.ic:
```diff
@@ -1,5 +1,5 @@
 # riscOVPsim configuration file converted from YAML
---variant RV32IMC
+--variant RV32IMCB
```
```suggestion
--variant RV32I
--override riscvOVPsim/cpu/add_Extensions=MCB
```

Structuring the supported extensions in this way is required for OVPsim to recognize bitmanip instructions and successfully run, otherwise it will still fail.

vogelpi replied:
:+1: thanks for the suggestion

udinator commented on dv/uvm/core_ibex/riscv_dv_extension/testlist.yaml:
```diff
@@ -587,3 +587,13 @@
+- test: riscv_bitmanip_test
+  desc: >
+    Random instruction test with supported B extension instructions
+  iterations: 1
```
nit: can you change this to 10 iterations?

vogelpi replied:
Yep, done.

vogelpi commented:
Thanks @udinator for checking out OVPsim, providing the required fix, and for clarifying the situation around current test failures.

I have implemented your suggestions, rebased on current master and enabled verification of Zbs for which the RTL got added in the meantime.

udinator commented:
@vogelpi - LGTM, approved!
Can you run one iteration of the new `riscv_bitmanip_test` with the new ZBS extension enabled just to make sure the flow works smoothly before you merge?

vogelpi commented:
Yes, it is working. Thanks for your approval @udinator !
